File: src/dominio/cpf.ts
/**
 * Máscara progressiva de CPF, aplicada a cada tecla: mantém só os dígitos
 * (no máximo 11) e devolve o quanto de 000.000.000-00 já cabe neles. Os
 * separadores nunca sobram no fim ("529", não "529."), então apagar de trás
 * para frente sempre apaga um dígito.
 */
export function aplicarMascaraCpf(entrada: string): string {
  const digitos = entrada.replace(/\D/g, "").slice(0, 11);
  const grupos = [digitos.slice(0, 3), digitos.slice(3, 6), digitos.slice(6, 9)]
    .filter((grupo) => grupo !== "")
    .join(".");
  const verificadores = digitos.slice(9);
  return verificadores === "" ? grupos : `${grupos}-${verificadores}`;
}

/** Aplica a máscara 000.000.000-00 quando a entrada tem 11 dígitos. */
export function formatarCpf(cpf: string): string {
  const digitos = cpf.replace(/\D/g, "");
  if (digitos.length !== 11) return cpf;
  return digitos.replace(/^(\d{3})(\d{3})(\d{3})(\d{2})$/, "$1.$2.$3-$4");
}

/** Valida os dígitos verificadores de um CPF, com ou sem máscara. */
export function validarCpf(cpf: string): boolean {
  const digitos = cpf.replace(/\D/g, "");
  if (digitos.length !== 11) return false;
  // Sequências repetidas (111.111.111-11 etc.) passam na aritmética dos
  // verificadores, mas são CPFs inválidos por definição.
  if (/^(\d)\1{10}$/.test(digitos)) return false;

  const verificador = (quantidade: number): number => {
    let soma = 0;
    for (let i = 0; i < quantidade; i++) {
      soma += Number(digitos[i]) * (quantidade + 1 - i);
    }
    const resto = soma % 11;
    return resto < 2 ? 0 : 11 - resto;
  };

  return (
    verificador(9) === Number(digitos[9]) &&
    verificador(10) === Number(digitos[10])
  );
}

File: src-tauri/src/cpf.rs
//! Validação de CPF no backend — espelho de `src/dominio/cpf.ts`.
//!
//! A rota de Auto-cadastro (ADR-0003) recebe cadastros de qualquer navegador
//! da rede local, então o backend valida o CPF por conta própria em vez de
//! confiar na SPA.

/// Valida os dígitos verificadores de um CPF, com ou sem máscara.
pub fn validar(cpf: &str) -> bool {
    let digitos: Vec<u32> = cpf.chars().filter_map(|c| c.to_digit(10)).collect();
    if digitos.len() != 11 {
        return false;
    }
    // Sequências repetidas (111.111.111-11 etc.) passam na aritmética dos
    // verificadores, mas são CPFs inválidos por definição.
    if digitos.iter().all(|&digito| digito == digitos[0]) {
        return false;
    }

    let verificador = |quantidade: usize| -> u32 {
        let soma: usize = digitos[..quantidade]
            .iter()
            .enumerate()
            .map(|(i, &digito)| digito as usize * (quantidade + 1 - i))
            .sum();
        let resto = soma % 11;
        if resto < 2 {
            0
        } else {
            (11 - resto) as u32
        }
    };

    verificador(9) == digitos[9] && verificador(10) == digitos[10]
}

#[cfg(test)]
mod testes {
    use super::validar;

    #[test]
    fn cpf_com_digitos_verificadores_corretos_e_valido() {
        assert!(validar("52998224725"));
    }

    #[test]
    fn mascara_nao_atrapalha_a_validacao() {
        assert!(validar("529.982.247-25"));
    }

    #[test]
    fn digito_verificador_errado_e_invalido() {
        assert!(!validar("52998224724"));
    }

    #[test]
    fn sequencia_repetida_e_invalida_mesmo_passando_na_aritmetica() {
        assert!(!validar("11111111111"));
    }

    #[test]
    fn tamanho_diferente_de_onze_digitos_e_invalido() {
        assert!(!validar("5299822472"));
        assert!(!validar("529982247255"));
        assert!(!validar(""));
    }
}
